File: 02/python/problem2.py
import itertools as it

# generator of even fibonacci numbers
def even_fibs():
    a, b = 0, 2
    while True:
        yield a
        a, b = b, 4 * b + a

# take all the even fibs that are less than 40 million and sum them
print(sum(it.takewhile(lambda x: x <= 40_000_000, even_fibs())))

File: 02/rust/problem2.rs
// an iterator of all the even fibs
fn even_fibs() -> impl Iterator<Item = usize> {
    let mut tup = (0, 2);
    std::iter::from_fn(move || {
        tup = (tup.1, 4 * tup.1 + tup.0);
        Some(tup.1)
    })
}

fn main() {
    // take the even fibs less than 40_000_000 and sum them, same as python
    println!("{}", even_fibs().take_while(|x| x <= &40_000_000).sum::<usize>())
}
